src/commands/restartServer.ts

```diff
@@ -0,0 +1,51 @@
+import {
+  ShutdownRequest,
+  ExitNotification
+} from "vscode-languageserver-protocol";
+import { LanguageClient } from "../interfaces/LanguageClient";
+import { exec, ChildProcess } from "child_process";
+import { ShowMessage, normalize } from "../interfaces/ShowMessage";
+
+const timeoutMs = 4000;
+
+/**
+ * Try to gracefully shutdown the server using LSP `shutdown` and `exit` endpoints.
+ * If the server is unresponsive, fall back to killing its process.
+ *
+ * @param client
+ * @param workspace
+ */
+export function restartServer(
```

Comment:
Not really relevant, but the code actions to change functions to async like it suggests for `restartServer` are so nice in typescript. #metalsGoals

Reply:
yeah, they work really well and also take care of changing stuff like .catch to try/catch in the function body. (I personally dislike async/await, but that’s another story)

I agree TS code actions are a gold standard we should look at closely!